kong/tools/emmy_debugger.lua

```diff
@@ -0,0 +1,79 @@
+local pl_path = require "pl.path"
+local utils = require "kong.tools.utils"
+
+local debugger = os.getenv("KONG_EMMY_DEBUGGER")
+local emmy_debugger_host = os.getenv("KONG_EMMY_DEBUGGER_HOST") or "localhost"
+local emmy_debugger_port = os.getenv("KONG_EMMY_DEBUGGER_PORT") or 9966
+local emmy_debugger_wait = os.getenv("KONG_EMMY_DEBUGGER_WAIT")
+local emmy_debugger_source_path = utils.split(os.getenv("KONG_EMMY_DEBUGGER_SOURCE_PATH") or "", ":")
+
+local function find_source(path)
+  if pl_path.exists(path) then
+    return path
+  end
+
+  if path:match("^=") then
+    -- code is executing from .conf file, don't attempt to map
+    return path
+  end
+
+  for _, source_path in ipairs(emmy_debugger_source_path) do
+    local full_path = pl_path.join(source_path, path)
+    if pl_path.exists(full_path) then
+      return full_path
+    end
+  end
+
+  ngx.log(ngx.ERR, "source file " .. path .. " not found in KONG_EMMY_DEBUGGER_SOURCE_PATH")
+
+  return path
+end
+
+local function init()
+  if not debugger then
+    return
+  end
+
+  if not pl_path.isabs(debugger) then
+    ngx.log(ngx.ERR, "KONG_EMMY_DEBUGGER (" .. debugger .. ") must be an absolute path")
+    return
+  end
+  if not pl_path.exists(debugger) then
+    ngx.log(ngx.ERR, "KONG_EMMY_DEBUGGER (" .. debugger .. ") file not found")
+    return
+  end
+  local ext = pl_path.extension(debugger)
+  if ext ~= ".so" and ext ~= ".dylib" then
+    ngx.log(ngx.ERR, "KONG_EMMY_DEBUGGER (" .. debugger .. ") must be a .so (Linux) or .dylib (macOS) file")
+    return
+  end
+  if ngx.worker.id() ~= 0 then
+    ngx.log(ngx.ERR, "KONG_EMMY_DEBUGGER is only supported in the first worker process, suggest setting KONG_NGINX_WORKER_PROCESSES to 1")
+    return
+  end
+
+  ngx.log(ngx.NOTICE, "loading EmmyLua debugger " .. debugger)
+
+  _G.emmy = {
+    fixPath = find_source
+  }
+
+  package.cpath = package.cpath .. ';' .. pl_path.dirname(debugger) .. '/?' .. ext
```

Comment:
Was it that there could be a special `;;` ending on these paths, is it preserved?

Reply:
I have been thinking about just this a Moment ago, from another angle:  We don't actually want to mess with `package.cpath` other than for loading the debugger, so I have changed the code to save the old value, set it to just the path of the debugger, `require` the debugger and then re-set `package.cpath` to what it was before.